packages/medusa/src/repositories/discount-condition.ts

```diff
@@ -108,39 +116,170 @@ export class DiscountConditionRepository extends Repository<DiscountCondition> {
   > {
     let toInsert: { condition_id: string; [x: string]: string }[] | [] = []
 
-    const { fromTable, resourceId } = this.getResourceIdentifiers(type)
+    const { joinTableName, joinTableForeignKey } =
+      this.getJoinTableResourceIdentifiers(type)
 
-    if (!fromTable || !resourceId) {
+    if (!joinTableName || !joinTableForeignKey) {
       return Promise.resolve([])
     }
 
     toInsert = resourceIds.map((pId) => ({
       condition_id: conditionId,
-      [resourceId]: pId,
+      [joinTableForeignKey]: pId,
     }))
 
     const insertResult = await this.createQueryBuilder()
       .insert()
       .orIgnore(true)
-      .into(fromTable)
+      .into(joinTableName)
       .values(toInsert)
       .execute()
 
     if (overrideExisting) {
       await this.createQueryBuilder()
         .delete()
-        .from(fromTable)
+        .from(joinTableName)
         .where({
           condition_id: conditionId,
-          [resourceId]: Not(In(resourceIds)),
+          [joinTableForeignKey]: Not(In(resourceIds)),
         })
         .execute()
     }
 
     return await this.manager
-      .createQueryBuilder(fromTable, "discon")
+      .createQueryBuilder(joinTableName, "discon")
       .select()
       .where(insertResult.identifiers)
       .getMany()
   }
+
+  async queryConditionTable({
+    type,
+    condId,
+    resourceId,
+  }): Promise<
+    (
+      | DiscountConditionProduct
+      | DiscountConditionProductType
+      | DiscountConditionProductCollection
+      | DiscountConditionProductTag
+      | DiscountConditionCustomerGroup
+    )[]
+  > {
+    const { joinTableName, joinTableForeignKey } =
+      this.getJoinTableResourceIdentifiers(type)
+
+    if (joinTableName) {
+      return await this.manager
+        .createQueryBuilder(joinTableName, "conds")
+        .select()
+        .where(`${"conds"}.${joinTableForeignKey} = :resourceId`, {
+          resourceId,
+        })
+        .andWhere(`${"conds"}.condition_id = :condId`, {
+          condId,
+        })
+        .getMany()
+    }
+
+    return []
+  }
+
+  async isValidForProduct(
+    discountRuleId: string,
+    product: Product
```

Comment:
Ideally, I think this should only get the product id. So that Postgres would take care of things like `for (const tag of product.tags)`.

A quick and dirty idea of the SQL would maybe be like this:

Tags
```
SELECT COUNT(UNIQUE dc.id) FROM "discount_condition_product_tag" dc
INNER JOIN "product_tag_product" ptp ON dc.tag_id = ptp.id AND product_id = {product_id}
WHERE dc.discount_id = {discount_id}
```

Type
```
SELECT COUNT(UNIQUE dc.id) FROM "discount_condition_product_type" dc
INNER JOIN "product" p ON dc.type_id = p.type_id
WHERE dc.discount_id = {discount_id}
```

Collection
```
SELECT COUNT(UNIQUE dc.id) FROM "discount_condition_product_collection" dc
INNER JOIN "product" p ON dc.collection_id = p.collection_id
WHERE dc.discount_id = {discount_id}
```

My primary argument for suggesting this is that we avoid having to remember to join tags before calling the method and potentially if new condition types are added in the future this will be more maintainable. Let me know what you think?

Reply:
In TypeORM terms it would prob look something like this:

```
const count = qb.innerJoin("product_tag_product", "ptp", "dc.product_tag_id = ptp.tag_id AND ptp.product_id = :productId", { productId: product_id }).getCount()
```

And then we can return valid status based on `count` and operator.